services/src/main/java/org/keycloak/services/managers/DefaultBruteForceProtector.java

```diff
@@ -261,125 +151,61 @@ protected void sendEvent(KeycloakSession session, RealmModel realm, UserLoginFai
+        UserLoginFailureModel userLoginFailure = getUserFailureModel(session, realm, userId);
+        if(userLoginFailure == null) return;
         if (logger.isDebugEnabled()) {
-            UserModel model = session.users().getUserById(getRealmModel(session, event), userId);
+            UserModel model = session.users().getUserById(realm, userId);
             logger.debugv("user {0} successfully logged in, clearing all failures", model.getUsername());
         }
-        user.clearFailures();
-    }
-
-    protected void logFailure(LoginEvent event) {
-        failures++;
-        long delta = 0;
-        if (lastFailure > 0) {
-            delta = Time.currentTimeMillis() - lastFailure;
-            if (delta > (long)maxDeltaTimeSeconds * 1000L) {
-                totalTime = 0;
-
-            } else {
-                totalTime += delta;
-            }
-        }
+        userLoginFailure.clearFailures();
     }
 
     @Override
     public void failedLogin(RealmModel realm, UserModel user, ClientConnection clientConnection) {
-        try {
-            FailedLogin event = new FailedLogin(realm.getId(), user.getId(), clientConnection);
-            queue.offer(event);
-            // wait a minimum of seconds for type to process so that a hacker
-            // cannot flood with failed logins and overwhelm the queue and not have notBefore updated to block next requests
-            // todo failure HTTP responses should be queued via async HTTP
-            event.latch.await(5, TimeUnit.SECONDS);
-        } catch (InterruptedException e) {
-        }
+        processLogin(realm, user, clientConnection, false);
+        // wait a minimum of seconds for type to process so that a hacker
+        // cannot flood with failed logins and overwhelm the queue and not have notBefore updated to block next requests
+        // todo failure HTTP responses should be queued via async HTTP
+        //event.latch.await(5, TimeUnit.SECONDS);
         logger.trace("sent failure event");
     }
 
     @Override
-    public void successfulLogin(final RealmModel realm, final UserModel user, final ClientConnection clientConnection) {
-        SuccessfulLogin event = new SuccessfulLogin(realm.getId(), user.getId(), clientConnection);
-        queue.offer(event);
+    public void successfulLogin(RealmModel realm, UserModel user, ClientConnection clientConnection) {
+        processLogin(realm, user, clientConnection, true);
         logger.trace("sent success event");
     }
 
+    private void processLogin(RealmModel realm, UserModel user, ClientConnection clientConnection, boolean success) {
+        KeycloakSession session = factory.create();
+        ExecutorsProvider provider = session.getProvider(ExecutorsProvider.class);
+        ExecutorService executor = provider.getExecutor("bruteforce");
+        executor.execute(() -> {
+            try {
+                session.getTransactionManager().begin();
```

Comment:
This is new thread and as such, it should probably also use dedicated `KeycloakSession` instance. Using same `KeycloakSession` within many threads can have bad side-effects as this is supposed to be per-thread object.

I think something like this should work (this will make sure to use dedicated KeycloakSession together with dedicated transaction and also automatically handles transaction rollbacks as well:)

```
executor.execute(() -> {
    KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), () -> {
        if(success) {
            success(session, realm, user.getId());
        } else {
            failure(session, realm, user.getId(), clientConnection.getRemoteAddr(), Time.currentTimeMillis());
        }
    });
});
```

Reply:
Done